src/iop/diffuse.c

```diff
@@ -0,0 +1,1481 @@
+#define DEBUG_DUMP_PFM 0
+
+DT_MODULE_INTROSPECTION(1, dt_iop_diffuse_params_t)
+
+#define MAX_NUM_SCALES 12
+typedef struct dt_iop_diffuse_params_t
+{
+  // global parameters
+  int iterations;           // $MIN: 1   $MAX: 128   $DEFAULT: 1  $DESCRIPTION: "iterations"
+  float sharpness;          // $MIN: -1.  $MAX: 1.   $DEFAULT: 0. $DESCRIPTION: "sharpness"
+  int radius;               // $MIN: 1   $MAX: 512   $DEFAULT: 8  $DESCRIPTION: "radius"
+  float regularization;     // $MIN: 0. $MAX: 4.   $DEFAULT: 0. $DESCRIPTION: "edge sensitivity"
+  float variance_threshold; // $MIN: -2. $MAX: 2.   $DEFAULT: 0. $DESCRIPTION: "edge threshold"
+
+  float anisotropy_first;         // $MIN: -10. $MAX: 10.   $DEFAULT: 0. $DESCRIPTION: "1st order anisotropy"
+  float anisotropy_second;        // $MIN: -10. $MAX: 10.   $DEFAULT: 0. $DESCRIPTION: "2nd order anisotropy"
+  float anisotropy_third;         // $MIN: -10. $MAX: 10.   $DEFAULT: 0. $DESCRIPTION: "3rd order anisotropy"
+  float anisotropy_fourth;        // $MIN: -10. $MAX: 10.   $DEFAULT: 0. $DESCRIPTION: "4th order anisotropy"
+
+  float threshold; // $MIN: 0.  $MAX: 8.   $DEFAULT: 0. $DESCRIPTION: "luminance masking threshold"
+
+  float first; // $MIN: -1. $MAX: 1.   $DEFAULT: 0. $DESCRIPTION: "1st order speed"
+  float second; // $MIN: -1. $MAX: 1.   $DEFAULT: 0. $DESCRIPTION: "2nd order speed"
+  float third; // $MIN: -1. $MAX: 1.   $DEFAULT: 0. $DESCRIPTION: "3rd order speed"
+  float fourth; // $MIN: -1. $MAX: 1.   $DEFAULT: 0. $DESCRIPTION: "4th order speed"
+} dt_iop_diffuse_params_t;
+
+
+typedef struct dt_iop_diffuse_gui_data_t
+{
+  GtkWidget *iterations, *fourth, *third, *second, *radius, *sharpness, *threshold, *regularization, *first,
+      *anisotropy_first, *anisotropy_second, *anisotropy_third, *anisotropy_fourth, *regularization_first, *variance_threshold;
+} dt_iop_diffuse_gui_data_t;
+
+typedef struct dt_iop_diffuse_global_data_t
+{
+  int kernel_wavelets_decompose;
+  int kernel_diffuse_build_mask;
+  int kernel_diffuse_inpaint_mask;
+  int kernel_diffuse_pde;
+} dt_iop_diffuse_global_data_t;
+
+
+// only copy params struct to avoid a commit_params()
+typedef struct dt_iop_diffuse_params_t dt_iop_diffuse_data_t;
+
+
+typedef enum dt_isotropy_t
+{
+  DT_ISOTROPY_ISOTROPE = 0, // diffuse in all directions with same intensity
+  DT_ISOTROPY_ISOPHOTE = 1, // diffuse more in the isophote direction (orthogonal to gradient)
+  DT_ISOTROPY_GRADIENT = 2  // diffuse more in the gradient direction
+} dt_isotropy_t;
+
+
+#ifdef _OPENMP
+#pragma omp declare simd
+#endif
+static inline dt_isotropy_t check_isotropy_mode(const float anisotropy)
+{
+  // user param is negative, positive or zero. The sign encodes the direction of diffusion, the magnitude encodes the ratio of anisotropy
+  // ultimately, the anisotropy factor needs to be positive before going into the exponential
+  if(anisotropy == 0.f) return DT_ISOTROPY_ISOTROPE;
```

Comment:
style: better having all return on a new line to have them aligned and more readable.

Reply:
done